Fix the logic in how we determine BuildConfiguration.current

AliSoftware opened the pull request:
See AINFRA-856

## Description

We can't rely on `#if ALPHA` anymore  in `BuildConfiguration.swift`'s code now that this file have been moved to `Modules/Sources/Experiments` as a SwiftPM module. This is because, in this context, SwiftPM does not apply any custom Build Conditional Settings from the Xcode project to the SwiftPM dependencies it builds (and instead the Modules are built with build settings that are independent from the consuming target / project using them)

So we have to rely on a different trick instead, which is to read the configuration from the `Info.plist` file of the app bundle.

### How

 - We added a new `BuildConfigurationName` key to the app's `Info.plist` and set its value to `$(CONFIGURATION)`, letting Xcode do the substitution during the pre-processing of the `Info.plist` to replace it with the build settings's value
 - The `BuildConfiguration.current` implementation now relies on reading the value from that new `Info.plist` key, instead of relying on `#if` conditions

### What's Next

> [!IMPORTANT]
> Note that `BuildConfiguration.swift` is probably not the only code in `Modules/` that rely on `#if ALPHA`. I've at least noticed `Modules/Sources/Yosemite/Stores/JustInTimeMessageStore.swift` is also relying on it, and will thus also likely be broken since this is compiled as a separate module by SwiftPM. You'll have to find a similar solution for this one too, and any potential other occurrences of `#if` conditions in `Modules/`.

## Steps to reproduce

Run the prototype build of a previous PR and observe that anything that relied on the `BuildConfiguration.current` used `.appStore` value

## Testing information

Run the prototype build of the current PR and validate that anything that relies on the `BuildConfiguration.current` use `.alpha` value and not `.appStore` anymore in this Prototype Build context.

> [!WARNING]
> On my end I have only tested this by running the `WooCommerce Alpha` in Xcode with a breakpoint, but haven't checked the Prototype Build on the PR (as I don't have the knowledge on how to test this on a WCiOS Prototype Build running in a device, as I'm assuming it involves setting up a proxy to intercept API requests or something like that where the `BuildConfiguration.current` value might appear somewhere? So cc @jaclync I'll let you test this more thoroughly

---
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~

AliSoftware committed:
Fix the logic in how we determine BuildConfiguration.current

We can't rely on `#if ALPHA` anymore now that `BuildConfiguration.swift` have been moved to `Modules/Sources/Experiments` as a SwiftPM module, because in this context SwiftPM does not apply any custom Build Conditional settings from the Xcode project to the SwiftPM dependencies it builds (and instead the Modules are built with build settings that are independent from the consuming target / project using them)

So we have to rely on a different trick instead, which is to read the configuration from the `Info.plist` file of the app bundle.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: PR is not assigned to a milest...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware committed:
Restore `#if DEBUG` around the `testingOverride` code

AliSoftware committed:
Force `Release-Alpha` configuration when building the Prototype Build

Even if someone changed the configuration associated with the `WooCommere Alpha` scheme in Xcode, this should enforce it back to `Release-Alpha` on CI at least

jaclync reviewed (approved):
LGTM! This seems like a good solution that works both locally and in CI. Tested locally printing out `BuildConfiguration.current`. In the prototype build, the issue with the local feature flag enabled for debug/alpha build configuration is also fixed.

I have a PR ready to update other existing usages of `ALPHA` preprocessor flag preprocessor https://github.com/woocommerce/woocommerce-ios/pull/15845 with the same approach for the watch app.
